pkg/fetcher/fetcher.go

```diff
@@ -280,13 +296,32 @@ func (u *URLFetcher) fetch(ctx context.Context, url *url.URL, reqHdrs *http.Head
 			}
 		}
 
-		err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s.", http.StatusText(u.StatusCode), u.StatusCode, url)
+		// for all other non-retryable client error, grab the error message if there is one (#5951)
+		msg, err := u.extractErrResponseMessage(res.Body)
+		if err != nil {
+			log.Debugf("Nonretryable client error in unconventional format: %s", err)
+			err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s", http.StatusText(u.StatusCode), u.StatusCode, url)
```

Comment:
yes, you're right. one thing i'm not sure about: the `Nonretryable client error` is in the 4xx http code category, and we are treating it differently from all other unexpected http error code. Do we need to give user this info in some way or we can just say `Unexpected http code` for all error codes?